src/Targets/Microsoft.Managed.DesignTime.targets

```diff
@@ -283,4 +278,15 @@
   <!-- This target collects all the resolved references that are used to actually compile. -->
   <Target Name="CollectResolvedCompilationReferencesDesignTime" DependsOnTargets="CompileDesignTime" Returns="@(ReferencePathWithRefAssemblies)" />
 
+  <!-- This target collects all up-to-date file additions. -->
+  <Target Name="CollectUpToDateInputsDesignTime" DependsOnTargets="CompileDesignTime" Returns="@(UpToDateCheckInput)">
+    <!-- If the project we're compiling has a ruleset, make sure we check it for up-to-date checks. -->
+    <ItemGroup Condition="'$(ResolvedCodeAnalysisRuleSet)' != ''">
+      <UpToDateCheckInput Include="$([System.IO.Path]::GetFullPath($(ResolvedCodeAnalysisRuleSet)))" />
```

Comment:
I'm still learning the system, so it's possible that there's a better way of doing this and/or I've gotten confused on things.

My understanding & experience is that if we root the rules in the project file, then they come from the evaluation, before targets are run. In this specific case, that's a problem because we need to run the ResolveCodeAnalysisRuleSet target to get the actual ruleset path that we're going to need to look at. (As a side note, I guess we could just duplicate the logic of that target in the up to date check, but I'm trying to avoid doing too much input-specific logic if I can help it.) Otherwise ResolvedCodeAnalysisRuleSet is just empty at evaluation time and no item is created.

This speaks to a more general problem in that in the old system, targets cannot add up-to-date check inputs or outputs, which is a major limitation.

Regardless, my understanding of the system is that we're not going to get a design-time build specifically for up-to-date checks. We subscribe to the rule data source, so we'll get notified when the general design-time build finishes. Before that, if you try and build, we just throw up our hands and say we don't know. So my understanding is that we're not adding any expense to the system (well, maybe the amount of running an additional target), but we are potentially going to lose some early builds that might have been considered up to date in the old system.

Is all that correct, or am I missing things?